README.md
# Marketfaux

For when you want to use Marketo forms, but not its API.

## Why not use Marketo's Forms API?

Marketo's Forms API is opinionated in terms of style and behavior. In some cases, those opinions are terrible and outdated. Additionally, sometimes you just want to break the mold and do something that would not normally be possible, such as utilizing a JavaScript framework like React.

## What Marketfaux does (and what it doesn't do)

Marketfaux just provides functions to send data as if you've filled out a standard Marketo form. That way, you can implement the form the way you want and still have your form's results passed on to Marketo.

Marketfaux, however, is not a drop-in replacement for the Marketo Forms API. It does not generate the form's HTML or validate input—those bits will have to be implemented by *you*.

## API

### Return type

Marketfaux uses [Wretch](https://github.com/elbywan/wretch) to send requests using the Fetch API. Please see their documentation to learn more about [error handling](https://github.com/elbywan/wretch#catchers) and [responses](https://github.com/elbywan/wretch#response-types).

All functions return Wretch's `ResponseChain` type.

### Base URL, Munchkin ID, and form ID

The first three parameters for each function are the same as in Marketo's `loadForm` function:
1. Base URL (ex. `"//app-sjqe.marketo.com"`)
2. Munchkin ID (ex. `"718-GIV-198"`)
3. Form ID (ex. `621`)

### Functions

#### sendForm(baseUrl: string, munchkinId: string, formId: string | number, form: HTMLFormElement | string): ResponseChain

Bundle up the data from a `<form />` element and submit it. Accepts either the element itself or a query selector for the element.

##### Examples

```js
Marketfaux.sendForm("//app-sjqe.marketo.com", "718-GIV-198", 621, "form#my-id")
```

```js
Marketfaux.sendForm("//app-sjqe.marketo.com", "718-GIV-198", 621, document.querySelector("form#my-id"))
```

#### sendData(baseUrl: string, munchkinId: string, formId: string | number, form: FormData | Record<string, any>): ResponseChain

Sends provided data which can either be a [`FormData`](https://developer.mozilla.org/en/docs/Web/API/FormData) object or a standard key-value object.

##### Bracketization

Marketfaux will automatically bracketize keys with multiple values to conform to PHP's naming conventions. If those keys already have bracket notation, they will not be bracketized.

###### Input
```json
{
  "key": ["value1", "value2"]
}
```

###### Output
```yaml
key[]: value1
key[]: value2
```

##### Examples

```js
Marketfaux.sendData("//app-sjqe.marketo.com", "718-GIV-198", 621, { key1: "value1", key2: "value2" })
```

```js
const data = new FormData()
data.append('key1', 'value1')
data.append('key2', 'value2')

Marketfaux.sendData("//app-sjqe.marketo.com", "718-GIV-198", 621, data)
```

#### send(baseUrl: string, munchkinId: string, formId: string | number, payload: HTMLFormElement | string | FormData | Record<string, any>): ResponseChain

Convenience method that will automatically use either `sendForm` or `sendData` based on the type of the provided payload.

## Examples

### On the page

```html
<form id="my-form">
  ...
</form>

<script src="/path/to/marketfaux.min.js"></script>
<script>
  var form = document.getElementById('my-form');
  form.addEventListener('submit', function (event) {
    event.preventDefault();
    Marketfaux.send("//app-sjqe.marketo.com", "718-GIV-198", 621, form)
      .badRequest(function () {
        // Handle 400 error
      })
      .timeout(function () {
        // Handle 408 error
      })
      .internalError(function () {
        // Handle 500 error
      })
      .json(function (result) {
        // Handle success
      })
  })
</script>
```

### In an app

```js
import { send } from 'marketfaux'

const formData = {
  // Key-value object
}

send("//app-sjqe.marketo.com", "718-GIV-198", 621, formData)
  .badRequest(() => {
    // Handle 400 error
  })
  .timeout(() => {
    // Handle 408 error
  })
  .internalError(() => {
    // Handle 500 error
  })
  .json((result) => {
    // Handle success
  })
```

## What type of errors should I expect?

Marketo is surprisingly loose when it comes to the data submitted to it.

It ***will*** throw errors if:
- your base URL, Munchkin ID, or form ID is invalid
  - 400 Bad Request error
- there are connectivity issues such as user being offline or Marketo being unreachable

It will ***not*** throw errors if you submit:
- an incompatible value (such as a random string when something in date format is expected)
  - The value for this field will stay as it was before submitting
- non-existent fields
  - Nothing happens
- real fields that are not added to the form
  - Populates that field with the provided value
- without required fields
  - Skips those fields


CHANGELOG.md
# 1.0.1 (2019-08-27)

## Fixes
- Missing declaration file (d23c8d99959a3b0273f349d389756b7ce833f669)

# 1.0.0 (2019-08-26)

Initial


src/marketfaux.ts
import Cookies from 'js-cookie'
import Url from 'url-parse'
import wretch from 'wretch'

function isFormElement (element: unknown): element is HTMLFormElement {
  return typeof element === 'object' && element != null && element.constructor.name === 'HTMLFormElement'
}

function dupes<T> (arr: T[]): T[] {
  const duplicates: T[] = []
  const originals: T[] = []
  for (const item of arr) {
    if (originals.indexOf(item) === -1) {
      originals.push(item)
    } else if (duplicates.indexOf(item) === -1) {
      duplicates.push(item)
    }
  }
  return duplicates
}

function bracketize (data: FormData): FormData {
  const res = new FormData()
  const duplicates = dupes([...data.keys()])
  for (const entry of data.entries()) {
    const [key, value] = entry
    if (duplicates.indexOf(key) !== -1 && !/\[\]$/.test(key)) {
      res.append(`${key}[]`, value)
    } else {
      res.append(key, value)
    }
  }
  return res
}

/**
 * Send Marketo form data.
 *
 * @example
 * sendData("//app-sjqe.marketo.com", "718-GIV-198", 621, { FirstName: "Billy", LastName: "Eyelash" })
 *
 * @see {@link https://npm.im/wretch|wretch}
 */
export function sendData (domain: string, munchkinId: string, formId: number | string, formData: Record<string, any> | FormData) {
  const { Munchkin } = window as any
  if (Munchkin) {
    try {
      Munchkin.createTrackingCookie(true);
    } catch (e) { }
  }
  const url = new Url(domain)
  url.set('pathname', '/index.php/leadCapture/save2')
  const trackingToken = Cookies.get('_mkto_trk')
  let req = wretch(url.href)
    .options({ mode: 'cors', credentials: 'same-origin' })
    .accept('application/json')
  if (formData instanceof FormData) {
    const newData = bracketize(formData)
    newData.append('formid', formId.toString())
    newData.append('munchkinId', munchkinId)
    newData.append('formVid', formId.toString())
    if (trackingToken != null) {
      newData.append('_mkt_trk', trackingToken)
    }
    newData.append('_mktoReferrer', window.location.href)
    req = req.body(newData)
  } else if (typeof formData === 'object') {
    req = req.formData({
      ...formData,
      formid: formId,
      munchkinId,
      formVid: formId,
      '_mkt_trk': trackingToken,
      '_mktoReferrer': window.location.href
    })
  } else {
    throw new Error('Invalid form data. Must either be a key-value object or an instance of FormData.')
  }
  return req.post()
}

/**
 * Send Marketo form data based on an HTML form element.
 *
 * @example
 * sendForm("//app-sjqe.marketo.com", "718-GIV-198", 621, document.querySelector('form#my-form'));
 * @example
 * sendForm("//app-sjqe.marketo.com", "718-GIV-198", 621, 'form#my-form');
 *
 * @see {@link https://npm.im/wretch|wretch}
 */
export function sendForm (domain: string, munchkinId: string, formId: number | string, form: HTMLFormElement | string) {
  const el = typeof form === 'string' ? document.querySelector(form) : form
  if (el == null || !isFormElement(el)) {
    throw new Error('Invalid form element')
  }
  const formData = new FormData(el)
  return sendData(domain, munchkinId, formId, formData)
}

/**
 * Send Marketo form data based on either an HTML form element or raw data.
 *
 * @example
 * send("//app-sjqe.marketo.com", "718-GIV-198", 621, document.querySelector('form#my-form'));
 * @example
 * send("//app-sjqe.marketo.com", "718-GIV-198", 621, { FirstName: "Billy", LastName: "Eyelash" }))
 *
 * @see {@link https://npm.im/wretch|wretch}
 */
export function send (domain: string, munchkinId: string, formId: number | string, payload: HTMLFormElement | string | FormData | Record<string, any>) {
  if (typeof payload === 'string' || isFormElement(payload)) {
    return sendForm(domain, munchkinId, formId, payload)
  } else if (payload instanceof FormData || typeof payload === 'object') {
    return sendData(domain, munchkinId, formId, payload)
  } else {
    throw new Error('Unknown payload type')
  }
}
